### PR TITLE
Add HLD for H/W capabilities fields in platform.json

### PR DESCRIPTION
**Pull request by ArunSaravananBalachandran**
Add H/W capabilities fields in platform.json

### CODE REVIEW

File: doc/platform_json_enhancement.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+A new set of `capabilities` fields are introduced in platform.json, for providing platform specific capablities on control and characteristics of the components.
+
+For each component's attribute, the defined `capabilities` fields are as follows:
+
+- "control" : A boolean, 'true' if the given attribute can be controlled from the NOS, 'false' otherwise.
+- Attribute specific fields:
+    - status led - "color" - A list of the supported colors.
+    - speed
+        - "minimum" - Minimum recommended fan speed that can be set.
```
Comment (keboliu) [inline doc/platform_json_enhancement.md:47]: Could you please elaborate on who will be the consumer of this info? Also as platform facts for the test cases?

Comment (ArunSaravananBalachandran) [reply doc/platform_json_enhancement.md:47]: Currently, these fields would be used as platform facts for platform API test suite.
But, these can also be consulted by any platform daemon before executing the respective APIs.

File: doc/platform_json_enhancement.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+Currently, `platform.json` is used for providing the expected structure of the platform components and interface details for supporting dynamic port breakout.
+
+### Platform capabilities field
+
+A new set of `capabilities` fields are introduced in platform.json, for providing platform specific capablities on control and characteristics of the components.
+
+For each component's attribute, the defined `capabilities` fields are as follows:
+
+- "control" : A boolean, 'true' if the given attribute can be controlled from the NOS, 'false' otherwise.
```
Comment (jleveque) [inline doc/platform_json_enhancement.md:43]: Although it might be a bit longer, maybe this key should be named `controllable` for clarity?

Also, is there a default value assumed if this is omitted? I assume we would default to `true`?

Comment (ArunSaravananBalachandran) [reply doc/platform_json_enhancement.md]: > Although it might be a bit longer, maybe this key should be named `controllable` for clarity?

As suggested, the name of the key has been changed from `control` to `controllable`

> Also, is there a default value assumed if this is omitted? I assume we would default to `true`?

For the default value, the application/test suite can assume it as `true` to maintain the current behavior

Comment (ArunSaravananBalachandran) [reply doc/platform_json_enhancement.md]: > Could you add a note about this in the doc, please? If `controllable` is not specified, it is assumed to be true.

Have included the default value for 'controllable'

File: doc/platform_json_enhancement.md (hunk 1 above)
Comment (jleveque) [reply doc/platform_json_enhancement.md:49]: @zzhiyuan: I'm not sure I understand your comment. The main usage of these values is for testing purposes to understand boundaries when testing the device. Can you please elaborate?

Comment (zzhiyuan) [reply doc/platform_json_enhancement.md:49]: Ah, I meant more for the latter part, "these can also be consulted by any platform daemon before executing the respective APIs."

Since there is already the fantray abstraction, I believe it is better for platform vendor to control the min/max fan speed which may differ by model, via thermal manager. Having values in platform.json for purpose of testing is fine.

File: doc/platform_json_enhancement.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+Currently, `platform.json` is used for providing the expected structure of the platform components and interface details for supporting dynamic port breakout.
+
+### Platform capabilities field
+
+A new set of `capabilities` fields are introduced in platform.json, for providing platform specific capablities on control and characteristics of the components.
+
+For each component's attribute, the defined `capabilities` fields are as follows:
+
+- "controllable" : A boolean, 'true' if the given attribute can be controlled from the NOS, 'false' otherwise.
```
Comment (zzhiyuan) [inline doc/platform_json_enhancement.md:43]: Can you elaborate in this context what controllable from the NOS mean? Most of these attributes would be managed by a daemon, either directly or via vendor code. So if say you change the fan speed, the thermalctld would change it back, and you'd have to stop the container in order to change it. Is this deemed controllable?

Comment (zzhiyuan) [reply doc/platform_json_enhancement.md:43]: Can you maybe reword to:

`- "controllable" : Boolean, 'true' if the given attribute can be controlled from the NOS, 'false' if it cannot. Defaults to 'true.'`

So you don't need the note below.

Comment (ArunSaravananBalachandran) [reply doc/platform_json_enhancement.md:43]: > Can you elaborate in this context what controllable from the NOS mean? Most of these attributes would be managed by a daemon, either directly or via vendor code. So if say you change the fan speed, the thermalctld would change it back, and you'd have to stop the container in order to change it. Is this deemed controllable?

If a given platform component can be controlled from the NOS (software-controlled), it is deemed to be 'controllable' and 
if a platform component is controlled by a dedicated platform controller (like BMC) and is not controllable from the NOS running on the CPU, it is deemed as 'not controllable'.

**Review by jleveque — APPROVED** [review]

File: doc/platform_json_enhancement.md (hunk 1 above)
Comment (jleveque) [reply doc/platform_json_enhancement.md:47]: Ah. Yes, I agree with you. Platform daemons really shouldn't be referring to these values.
